benchmark/profile_pipeline_api.py

```diff
@@ -0,0 +1,252 @@
+                data['conversations'][1]['value']) for data in dataset]
+
+    # pre-sample to avoid go through all the dataset
+    dataset = random.sample(dataset, max(int(num_requests * 1.2), 1000))
+
+    # Tokenize the prompts and completions.
+    prompts = [prompt for prompt, _ in dataset]
+    prompt_token_ids = tokenizer(prompts).input_ids
+    completions = [completion for _, completion in dataset]
+    completion_token_ids = tokenizer(completions).input_ids
+    tokenized_dataset = []
+    for i in range(len(dataset)):
+        output_len = len(completion_token_ids[i])
+        tokenized_dataset.append((prompts[i], prompt_token_ids[i], output_len))
+
+    # Filter out too long sequences.
+    filtered_dataset: List[Tuple[str, int, int]] = []
+    for prompt, prompt_token_ids, output_len in tokenized_dataset:
+        prompt_len = len(prompt_token_ids)
+        if prompt_len < 4 or output_len < 4:
+            # Prune too short sequences.
+            continue
+        if prompt_len > 1024 or prompt_len + output_len > 2048:
+            # Prune too long sequences.
+            continue
+        filtered_dataset.append((prompt, prompt_len, output_len))
+
+    # Sample the requests.
+    sampled_requests = random.sample(filtered_dataset, num_requests)
+    return sampled_requests
+
+
+class Engine:
+
+    def __init__(self, model_path: str, engine_config, csv: str):
+        self.pipe = pipeline(model_path, backend_config=engine_config)
+        self.tokenizer = AutoTokenizer.from_pretrained(model_path,
+                                                       trust_remote_code=True)
+
+        self.csv = csv
+        self.pbar = None
+
+    def process_request(self, requests, concurrency, temperature, top_p, top_k,
+                        stream_output):
+
+        stats = OrderedDict(
+            (session_id, None) for session_id in range(len(requests)))
+        prompts = [prompt for prompt, _, _ in requests]
+        gen_configs = [
+            GenerationConfig(temperature=temperature,
+                             top_p=top_p,
+                             top_k=top_k,
+                             ignore_eos=True,
+                             max_new_tokens=output_len)
+            for _, _, output_len in requests
+        ]
+
+        start = time.perf_counter()
+        if stream_output:
+            self.pbar = tqdm(total=len(requests))
+            for output in self.pipe.stream_infer(prompts,
+                                                 gen_configs,
+                                                 do_preprocess=False):
+                session_id = output.session_id
+                n_token = output.generate_token_len
+                finish_reason = output.finish_reason
+                stats[session_id] = (n_token, finish_reason)
+                if finish_reason is not None:
+                    self.pbar.update(1)
+        else:
+            for output in self.pipe(prompts, gen_configs, do_preprocess=False):
```

Comment:
hope we have progress bar here.

Reply:
progress bar can only be add inside `batch_infer`.